### PR TITLE
Items are now added to sets using proper __hash__ function calls.

### PR DESCRIPTION
**Pull request by mcon**
Unsure what the point of adding items to the set in the `new_set` function in in `pyobject.rs`, given `set_new` in `objset.rs` does the same.

### CODE REVIEW

File: vm/src/frame.rs [hunk 1]
```diff
@@ -230,8 +230,8 @@ impl Frame {
                 Ok(None)
             }
             bytecode::Instruction::BuildSet { size, unpack } => {
-                let elements = self.get_elements(vm, *size, *unpack)?;
-                let py_obj = vm.ctx.new_set(elements);
+                let _elements = self.get_elements(vm, *size, *unpack)?;
+                let py_obj = vm.ctx.new_set();
```
Comment (windelbouwman) [inline vm/src/frame.rs]: We need to pass in the elements here when using set literals like `{1,2,3}`. Please revert this change.

Comment (windelbouwman) [reply vm/src/frame.rs]: Hmm, in retrospect it might be better to loop over the elements and call the set_add function with the vm, since you are right that the `__hash__` function requires the vm.

File: vm/src/pyobject.rs [hunk 2]
```diff
@@ -871,7 +873,7 @@ pub enum PyObjectPayload {
         elements: objdict::DictContentType,
     },
     Set {
-        elements: HashMap<usize, PyObjectRef>,
+        elements: HashMap<BigInt, PyObjectRef>,
```
Comment (windelbouwman) [inline vm/src/pyobject.rs]: Why change the hash result to `BigInt`? I thought that a hash always produces a native sized unsigned integer like `usize`. Is there a reason for using `BigInt` instead?

File: vm/src/obj/objset.rs [hunk 3]
```diff
@@ -3,32 +3,67 @@
...
+    item: &PyObjectRef,
+    f: &Fn(
+        &mut VirtualMachine,
+        &mut HashMap<BigInt, PyObjectRef>,
+        BigInt,
+        &PyObjectRef,
+    ) -> PyResult,
+) -> PyResult {
+    let hash_result: PyResult = vm.call_method(item, "__hash__", vec![]);
+    match hash_result {
```
Comment (windelbouwman) [inline vm/src/obj/objset.rs]: Instead of using hash_result in a match, use the `?` operator like this:
`let hash = vm.call_method(item, "__hash__", vec![])?;`
Next, use `vm.isinstance` to check the type of the python object and `objint::getvalue` to extract its payload data (the bigint).

**Comment by windelbouwman** [comment]
Thank you very much for your contribution! I made some remarks and had some questions on your changes.

File: vm/src/frame.rs (hunk 1 above)
Comment (mcon) [reply vm/src/frame.rs]: Sounds good to me - might me missing why we need to do this, when it looks like I can initialize a set from an iterable without the change you suggest

File: vm/src/pyobject.rs (hunk 2 above)
Comment (mcon) [reply vm/src/pyobject.rs]: __hash__ returns a python `int` which maps to `BigInt` in Rust-land :)

File: vm/src/obj/objset.rs (hunk 3 above)
Comment (mcon) [reply vm/src/obj/objset.rs]: Happy to change it if that's the convention for the project.

File: vm/src/pyobject.rs (hunk 2 above)
Comment (windelbouwman) [reply vm/src/pyobject.rs]: Yes, this is certainly true, but the range of this BigInt should fall into a usize category, following that most hashes are calculated using rust hashers which return usize. But then again, we could have python hashers which return larger numbers, so I guess you have a valid point!

File: vm/src/frame.rs (hunk 1 above)
Comment (windelbouwman) [reply vm/src/frame.rs]: Okay, we can initialize from iterable, this would be best, but we require the usage of the `elements` in some way in order to construct set literals.

File: vm/src/obj/objset.rs [hunk 4]
```diff
@@ -3,32 +3,67 @@
...
-pub fn sequence_to_hashmap(iterable: &Vec<PyObjectRef>) -> HashMap<usize, PyObjectRef> {
-    let mut elements = HashMap::new();
-    for item in iterable {
-        let key = item.get_id();
-        elements.insert(key, item.clone());
+fn perform_action_with_hash(
+    vm: &mut VirtualMachine,
+    elements: &mut HashMap<BigInt, PyObjectRef>,
+    item: &PyObjectRef,
+    f: &Fn(
```
Comment (windelbouwman) [inline vm/src/obj/objset.rs]: Is there a way to rewrite this code such that this function pass can be avoided? In my opinion creating a local function and passing that function is very powerful, but might be harder to understand what is going on.

File: vm/src/pyobject.rs (hunk 2 above)
Comment (adrian17) [reply vm/src/pyobject.rs]: https://github.com/python/cpython/blob/master/Objects/typeobject.c#L6379

if a returned Python `int` is out of range, it gets hashed by builtin `int.__hash__` hashing function. Thus, the result should always fit in an `usize`.

Comment (windelbouwman) [reply vm/src/pyobject.rs]: Maybe we can take the hash of the bigint and use that? That should be a `usize` thingy.

Comment (mcon) [reply vm/src/pyobject.rs]: Looks like we could do something like this: https://stackoverflow.com/questions/50484012/can-a-biginteger-be-truncated-to-an-i32-in-rust
I'm fairly busy for the next few days - but there's a chance I might manage to do this on the train tomorrow.

**Comment by mcon** [comment]
Should be good to go now: we're hashing every time for keys in the hashmap backing the set, when we could get away without for things smaller than `u64`, but I think we can consider that a future optimization.

File: vm/src/obj/objset.rs [hunk 5]
```diff
@@ -39,12 +64,40 @@ fn set_add(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
     );
     let mut mut_obj = s.borrow_mut();
 
-    if let PyObjectPayload::Set { ref mut elements } = mut_obj.payload {
-        let key = item.get_id();
-        elements.insert(key, item.clone());
-        Ok(vm.get_none())
-    } else {
-        Err(vm.new_type_error("set.add is called with no list".to_string()))
+    match mut_obj.payload {
```
Comment (windelbouwman) [inline vm/src/obj/objset.rs]: The `if let` syntax is better in this case of a match with a single arm and a catch all. At least, this is what clippy suggests.

**Comment by windelbouwman** [comment]
Great job! Thank you for your contribution to this project!
